README.md
Color Hexdump
=============

I got tired of hexdump being not in color. Now it's in color.

Other docs
----------

-   [INSTALL][install]
-   [NEWS][news]

Less helpful, but here for the sake of being here:

-   [AUTHORS][authors]
-   [COPYRIGHT][copyright]
-   [LICENSE][license]

Screen shot?
------------

Here's what it looks like:

    chexdump pdfa.pdf | more -f

![chexdump pdfa.pdf][pdfa]

This is a PDF where some binary data showed up. For comparison, the
output from `hexdump -C` for the same bytes looks like this:

    00000180  37 2f 52 6f 6f 74 20 31  36 20 30 20 52 2f 49 6e  |7/Root 16 0 R/In|
    00000190  66 6f 20 31 34 20 30 20  52 2f 49 44 5b 3c 31 34  |fo 14 0 R/ID[<14|
    000001a0  33 36 35 35 34 38 45 41  31 37 31 31 38 46 39 42  |365548EA17118F9B|
    000001b0  33 34 32 32 44 46 31 33  36 33 42 32 45 42 3e 3c  |3422DF1363B2EB><|
    000001c0  38 34 37 41 35 42 33 42  30 38 36 43 34 33 33 44  |847A5B3B086C433D|
    000001d0  38 37 35 31 36 45 45 42  38 34 36 38 45 37 34 45  |87516EEB8468E74E|
    000001e0  3e 5d 2f 50 72 65 76 20  34 33 35 38 35 34 37 3e  |>]/Prev 4358547>|
    000001f0  3e 0d 73 74 61 72 74 78  72 65 66 0d 30 0d 25 25  |>.startxref.0.%%|
    00000200  45 4f 46 0d 20 20 20 20  20 20 20 20 20 20 20 20  |EOF.            |
    00000210  20 20 20 20 20 20 20 0d  32 36 20 30 20 6f 62 6a  |       .26 0 obj|
    00000220  0d 3c 3c 2f 46 69 6c 74  65 72 2f 46 6c 61 74 65  |.<</Filter/Flate|
    00000230  44 65 63 6f 64 65 2f 49  20 31 33 33 2f 4c 65 6e  |Decode/I 133/Len|
    00000240  67 74 68 20 31 30 37 2f  4f 20 31 31 37 2f 53 20  |gth 107/O 117/S |
    00000250  36 39 3e 3e 73 74 72 65  61 6d 0d 0a 68 de 62 60  |69>>stream..h.b`|
    00000260  60 60 66 60 60 aa 06 92  ac 91 a7 18 c4 18 10 40  |``f``..........@|
    00000270  0c 28 c6 cc c0 c2 c0 71  10 c8 61 e2 9b 59 12 fa  |.(.....q..a..Y..|
    00000280  af 61 12 90 f5 54 2f 80  29 03 04 e1 a2 48 9a 38  |.a...T/.)....H.8|
    00000290  a1 98 81 41 99 41 98 41  fa 00 90 e1 c6 c0 f2 80  |...A.A.A........|
    000002a0  81 cf ee 4d 83 20 03 03  7f b7 ec 03 a8 3a a7 c0  |...M. .......:..|
    000002b0  2c 20 cd 08 c4 4a 40 cc  c7 e0 d4 60 07 e5 4f 01  |, ...J@....`..O.|
    000002c0  08 30 00 db 70 14 39 0d  65 6e 64 73 74 72 65 61  |.0..p.9.endstrea|
    000002d0  6d 0d 65 6e 64 6f 62 6a  0d 31 36 20 30 20 6f 62  |m.endobj.16 0 ob|
    000002e0  6a 0d 3c 3c 2f 4d 65 74  61 64 61 74 61 20 31 31  |j.<</Metadata 11|
    000002f0  20 30 20 52 2f 4f 75 74  6c 69 6e 65 73 20 39 20  | 0 R/Outlines 9 |
    00000300  30 20 52 2f 4f 75 74 70  75 74 49 6e 74 65 6e 74  |0 R/OutputIntent|
    00000310  73 20 31 32 20 30 20 52  2f 50 61 67 65 73 20 31  |s 12 0 R/Pages 1|
    00000320  30 20 30 20 52 2f 54 79  70 65 2f 43 61 74 61 6c  |0 0 R/Type/Catal|
    00000330  6f 67 3e 3e 0d 65 6e 64  6f 62 6a 0d 31 37 20 30  |og>>.endobj.17 0|
    00000340  20 6f 62 6a 0d 3c 3c 2f  43 6f 6e 74 65 6e 74 73  | obj.<</Contents|
    00000350  20 31 39 20 30 20 52 2f  43 72 6f 70 42 6f 78 5b  | 19 0 R/CropBox[|
    00000360  30 2e 30 20 30 2e 30 20  33 39 36 2e 30 20 36 31  |0.0 0.0 396.0 61|
    00000370  32 2e 30 5d 2f 4d 65 64  69 61 42 6f 78 5b 30 2e  |2.0]/MediaBox[0.|
    00000380  30 20 30 2e 30 20 33 39  36 2e 30 20 36 31 32 2e  |0 0.0 396.0 612.|
    00000390  30 5d 2f 50 61 72 65 6e  74 20 31 30 20 30 20 52  |0]/Parent 10 0 R|
    000003a0  2f 52 65 73 6f 75 72 63  65 73 3c 3c 2f 43 6f 6c  |/Resources<</Col|
    000003b0  6f 72 53 70 61 63 65 3c  3c 2f 44 65 66 61 75 6c  |orSpace<</Defaul|

You can also set skip (-s) and length (-n) just like in regular hexdump
if you want just pieces of files.

more -f
-------

Less and more don't naturally handle all the colors very well so if
you're piping to those, you'll want `less -R` and `more -f`.

Why is this in C instead of [your favorite language]
----------------------------------------------------

The most important answer here is that I like C.

But if it makes you feel better, I ran some benchmarking on really big
files, and this C version is about ten times faster than the first draft
I mocked up in Python.

Also I mean it's just code and not very much of it. I wrote a lot of
comments. Take a look; you can probably figure it out if you need to.

To do
-----

1.  Improve error handling; segfaulting just because a user gave a
    nonnumeric argument to `-n` or `-s` is unacceptable.

2.  I guess I could make it possible for you to configure your own
    colors if you don't like my totally awesome choices for some reason.

3.  If I can come up with a reason for `~/.config/chexdump` to exist, I
    guess I'll make that happen. Maybe that's how that configuration
    could work.

[authors]:    AUTHORS.md
[copyright]:  COPYRIGHT.md
[install]:    INSTALL.md
[license]:    LICENSE.txt
[news]:       NEWS.md
[pdfa]:       doc/pdfa.png


NEWS.md
chexdump 1.2 Release Notes
==========================

Added the beerware license. Now there is a licence, wow

chexdump 1.1 Release Notes
==========================

Now it's configurable using autoconf stuff, so you don't have to do
anything weird or special to make it work!


hexprint.h
#ifndef HEXPRINT_H
#define HEXPRINT_H

#include <stdio.h>
#include <string.h>

/* This is kinda an enum. */
#define COLOR_NORMAL  0
#define COLOR_CONTROL 1
#define COLOR_UTFLOW  2
#define COLOR_UTFHIGH 3

/* Sixteen bytes per line. */
#define BYTES_PER_LINE 0x10

/**
 * Print line
 *
 * Prints a row of hex bytes to the output stream.
 *
 * stream     - the output stream
 * pos        - the byte position in the input stream
 * bytes      - the size-16 array of bytes
 * stop       - integer [1,16] for where to stop printing
 * endstring  - char array of size at least 2 * (BYTES_PER_LINE + 1)
 */
void print_line (FILE*          stream,
                 unsigned int   pos,
                 unsigned char* bytes,
                 unsigned char  stop,
                 char*          endstring);

/**
 * Write byte
 *
 * Write a single byte to output and store an ASCII version of it in a
 * given string.
 *
 * stream     - the output stream
 * byte       - the actual byte
 * color      - the current output color
 * endstring  - char array of size at least 2 * (BYTES_PER_LINE + 1)
 * endlen     - current length of endstring
 */
void write_byte (FILE*          stream,
                 unsigned char  byte,
                 unsigned char* color,
                 char*          endstring,
                 unsigned char* endlen);

#endif /* HEXPRINT_H */


src/main.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <localconf.h>
#include "hexprint.h"

/* This is the help text. It can be used in printf or fprintf. This does
 * count on the presence of argv wherever this is used.
 */
#define USAGE_TEXT \
  "Usage: %s [-h | -v] [-n length] [-s skip] [file ...]\n", argv[0]

/* These can be returned by the argument parser to tell the main program
 * not to actually do anything.
 */
#define HALT_BUT_RETURN_ZERO  -1
#define HALT_AND_RETURN_ONE   -2

/* These are allocated in unistd. */
extern char *optarg;
extern int  optind;

/**
 * Parse arguments
 *
 * argc   - total argument count
 * argv   - full array of arguments
 * length - reference for desired length
 * offset - reference for desired offset
 *
 * Reads each argument looking for options. Sets parameters accordingly
 * and reorders the arguments such that the real ones are at the end of
 * argv.
 *
 * Returns a count of remaining arguments and sets optind to the index
 * of the first nonoption argument in argv.
 *
 * Negative return values are error codes:
 *
 * - HALT_BUT_RETURN_ZERO
 * - HALT_AND_RETURN_ONE
 */
int parse_args (int       argc,
                char*     argv[ ],
                long int* length,
                long int* offset);

/**
 * Read file object
 *
 * the_file - pointer to the FILE object (should be mode "rb")
 * length   - how many bytes to read from the file
 * offset   - where to start in the file
 *
 * This reads the FILE pointer you give it and outputs to stdout.
 */
int read_file_obj (FILE*    the_file,
                   long int length,
                   long int offset);

/**
 * Read one file
 *
 * path_to_file - string path to the file we want to read
 * length       - how many bytes to read from the file
 * offset       - where to start in the file
 *
 * This is just like reading a file object except that it opens the file
 * first and then closes it after. Rather than taking in an object, it
 * takes in a path to the file you want to read.
 */
int read_one_file (char*    path_to_file,
                   long int length,
                   long int offset);

/**
 * Read multiple files
 *
 * argc     - integer count of items in argv
 * argv     - array (size argc) of strings; the first will be ignored
 * start_i  - where to start in the array of strings
 * length   - how many bytes to read from the file
 * offset   - where to start in the file
 *
 * This takes in a file count as well as an array (sized to that count)
 * of file path strings. It outputs the filename of each before running
 * read_one_file on them.
 *
 */
int read_multiple_files (int      argc,
                         char*    argv[ ],
                         int      start_i,
                         long int length,
                         long int offset);

int main (int argc, char* argv[ ])
{
  int       arg_count;
  long int  length, offset;


  /* Now that we're done with getopt, we can get the count of nonoption
   * arguments.
   */
  arg_count = parse_args(argc, argv, &length, &offset);

  switch(arg_count)
  {
    case HALT_BUT_RETURN_ZERO:
      /* The user must have asked for help or something. We're exiting
       * before doing anything, but we're exiting clean.
       */
      return 0;

    case HALT_AND_RETURN_ONE:
      /* Something went wrong. */
      return 1;

    case 0:
      /* We've been given no arguments; read from stdin. */
      return read_file_obj(stdin, length, offset);

    case 1:
      /* We've been given one argument; read that file. */
      return read_one_file(argv[optind], length, offset);

    default:
      /* We've been given multiple arguments; read each file. */
      return read_multiple_files(argc, argv, optind, length, offset);
  }
}

int parse_args (int       argc,
                char*     argv[ ],
                long int* length,
                long int* offset)
{
  int option;

  /* Default these two to zero. Since a length of zero would be
   * nonsensical, we use it to mean "keep reading until EOF." An offset
   * of zero simply points us at the beginning of the file.
   */
  *length = 0;
  *offset = 0;

  while ((option = getopt(argc, argv, "hvn:s:")) != -1)
  {
    /* We've run getopt, and it's recorded an option. What happens next
     * depends on what that option is.
     */
    switch (option)
    {
      case 'h':
        /* Print help text and exit. */
        printf(USAGE_TEXT);
        return HALT_BUT_RETURN_ZERO;

      case 'v':
        /* Print the package string and exit. */
        printf("%s\n", PACKAGE_STRING);
        return HALT_BUT_RETURN_ZERO;

        /* This is for any options that just get loaded right into
         * integer variables. Never write anything twice!
         */
#define OPT_ATOI(CHAR, VAR)   \
      case CHAR:              \
        *VAR = atoi(optarg);  \
        break

        /* These options are just integers. */
        OPT_ATOI('n', length);
        OPT_ATOI('s', offset);

        /* Cleaning up after myself. */
#undef OPT_ATOI

      default:
        /* Print the usage text and error out. */
        fprintf(stderr, USAGE_TEXT);
        return HALT_AND_RETURN_ONE;
    }
  }

  /* Be sure we don't return any negative numbers. */
  if (optind > argc)  return 0;
  else                return argc - optind;
}

int read_file_obj (FILE*    the_file,
                   long int length,
                   long int offset)
{
  /* Just initializing these here. You'll see what they're about. */
  unsigned char bytes [BYTES_PER_LINE];
  size_t        linelen, bytesize;
  unsigned int  bytes_read, i;

  /* The dummy string is to avoid initializing the same block of bytes
   * over and over and over and etc.
   */
  char          dummy_string [2*(BYTES_PER_LINE + 1)];

  /* One byte per, uh, byte. */
  bytesize    = 1;

  /* We'll start at the offset. */
  bytes_read  = offset;

  if (offset != 0)
  {
    /* If we were given an offset greater than zero, we may have to read
     * an extra line at the beginning. We'll start by skipping to
     * wherever we're to begin in the file.
     */
    fseek(the_file, offset, SEEK_SET);

    /* Figure out whether we'll need to read a weird line. */
    i = offset % BYTES_PER_LINE;

    if (i != 0)
    {
      /* Read into the array, but only into the end of the array. */
      linelen = fread(bytes + i,
                      bytesize,
                      BYTES_PER_LINE - i,
                      the_file);

      if (length != 0 && length < linelen)
      {
        /* If we've already read too much, rewind. */
        fseek(the_file, length - linelen, SEEK_CUR);

        /* Set the linelen to something more appropriate. */
        linelen = length;
      }

      /* Print this piece of the line. */
      print_line(stdout, offset, bytes, linelen + i, dummy_string);

      /* Add to our byte count. */
      bytes_read += linelen;
    }
  }

  if (length == 0)
  {
    /* If there's no length, we can just read to the end. */
    while (!feof(the_file))
    {
#define READ_FULL_LINE(LENGTH)                                      \
      /* Read up to sixteen bytes. */                               \
      linelen = fread(bytes, bytesize, LENGTH, the_file);   \
                                                                    \
      /* Print this line. */                                        \
      print_line(stdout, bytes_read, bytes, linelen, dummy_string); \
                                                                    \
      /* Keep track of how many bytes we've read so far. */         \
      bytes_read += linelen

      /* Do the above stuff for real now. */
      READ_FULL_LINE(BYTES_PER_LINE);
    }
  }

  else
  {
    /* Otherwise, we were given a length. Add the offset to get a
     * stopping point.
     */
    length += offset;

    /* Repurpose the offset to point to the value bytes_read will be
     * when it's time to read our last line.
     */
    offset = (length / BYTES_PER_LINE) * BYTES_PER_LINE;

    /* Last, we'll set our length to the length of the final line. */
    length -= offset;

    while (bytes_read < offset && !feof(the_file))
    {
      /* Same as with no length limit, since we know we won't reach it
       * this time around.
       */
      READ_FULL_LINE(BYTES_PER_LINE);
    }

    if (length > 0 && !feof(the_file))
    {
      /* This is the last line. Instead of reading as much as possible,
       * read only up to our length max.
       */
      READ_FULL_LINE(length);

      /* (cleaning up) */
#undef READ_FULL_LINE
    }
  }

  /* Lastly, print the total. */
  printf("%08x\n", bytes_read);

  return 0;
}

int read_one_file (char*    path_to_file,
                   long int length,
                   long int offset)
{
  int   result;
  FILE  *the_file;

  if (strcmp(path_to_file, "-") == 0)
    /* The hyphen just means we read from stdin. */
    return read_file_obj(stdin, length, offset);

  /* Open the file. */
  the_file = fopen(path_to_file, "rb");

  /* Dump it. */
  result = read_file_obj(the_file, length, offset);

  /* Close it. */
  fclose(the_file);

  return result;
}

int read_multiple_files (int      argc,
                         char*    argv[ ],
                         int      start_i,
                         long int length,
                         long int offset)
{
  int i, result;

  i = start_i;

  /* Print the first filename, then the file. */
  printf("%s:\n", argv[i]);
  result = read_one_file(argv[i], length, offset);

  /* We also test for a nonzero result in each iteration. This way we
   * can immediately exit if needed.
   */
  for (i += 1; i < argc && result == 0; ++i)
  {
    /* Do the same with the rest of the files, but also add an extra
     * newline between them.
     */
    printf("\n%s:\n", argv[i]);
    result = read_one_file(argv[i], length, offset);
  }

  return result;
}

/* Cleaning up after myself is all. */
#undef USAGE_TEXT
#undef HALT_BUT_RETURN_ZERO
#undef HALT_AND_RETURN_ONE


hexprint.c
#include "hexprint.h"

#define COLOR_CODE(CODE) "\033[" CODE "m"
const char *COLOR_CODES[] = { COLOR_CODE("0"),      /* Normal */
                              COLOR_CODE("0;34"),   /* Blue */
                              COLOR_CODE("0;33"),   /* Yellow */
                              COLOR_CODE("0;35")};  /* Purple */
#undef COLOR_CODE

void print_line (FILE*          stream,
                 unsigned int   pos,
                 unsigned char* bytes,
                 unsigned char  stop,
                 char*          endstring)
{
  unsigned char i, start, column_stop, endlen, color;

  /* We initialize i and endlen to zero. Our color begins at normal.
   * We'll be starting in particular at the file position modulo the
   * number of bytes per line (probably zero). Lastly, we stop at the
   * halfway point along the line to add an extra space.
   */
  i           = 0;
  color       = COLOR_NORMAL;
  endlen      = 0;
  start       = pos % BYTES_PER_LINE;
  column_stop = BYTES_PER_LINE / 2;

  if (start >= stop)
    /* If there's not gonna be any data on this line, then don't print
     * anything.
     */
    return;

  if (column_stop > stop)
    /* If we're actually supposed to stop before the second column, then
     * we should stop writing the first column there as well.
     */
    column_stop = stop;

  /* We begin with our position. */
  fprintf(stream, "%08x  ", pos);

  if (start >= column_stop)
  {
    /* If we don't start until the second half of the line, we only need
     * to print a buncha spaces for this first half.
     */
    for (i; i < column_stop; ++i)
    {
      /* For each byte, print the requisite number of spaces. */
      fputs("   ", stream);
      endstring[endlen++] = ' ';
    }
  }

  else
  {
    /* We are starting in the first column, but we might not be starting
     * at zero. If we are starting at zero, this loop will never run.
     */
    for (i; i < start; ++i)
    {
      /* Print all the spaces corresponding to unreported bytes. */
      fputs("   ", stream);
      endstring[endlen++] = ' ';
    }

    /* We now begin printing actual data for sure. Add a pipe to our end
     * string.
     */
    endstring[endlen++] = '|';

    for (i; i < column_stop; ++i)
      /* Write each byte of data. */
      write_byte(stream, bytes[i], &color, endstring, &endlen);
  }

  /* The first column is done. Print a space. */
  fputs(" ", stream);

  if (i < stop)
  {
    /* We still have data to print. */
    if (i <= start)
    {
      /* Actually, we haven't even started printing data yet. Print
       * however much space we still need, if any.
       */
      for (i; i < start; ++i)
      {
        fputs("   ", stream);
        endstring[endlen++] = ' ';
      }

      /* Finally it's time to start printing data. Add the pipe. */
      endstring[endlen++] = '|';
    }

    for (i; i < stop; ++i)
      /* Write each byte of data. */
      write_byte(stream, bytes[i], &color, endstring, &endlen);
  }

  /* We're done printing actual data; print however many extra sets of
   * three spaces it'll take to align stuff right.
   */
  for (i; i < BYTES_PER_LINE; ++i)
    fputs("   ", stream);
  fputs(" ", stream);

  if (color != COLOR_NORMAL)
  {
    /* If we're currently on some weird color, go back to normal. */
    fputs(COLOR_CODES[COLOR_NORMAL], stream);
    endstring[endlen++] = COLOR_NORMAL;
  }

  for (i = 0; i < endlen; ++i)
  {
    /* Iterate through each character in the end-string. */
    if (endstring[i] < ' ')
      /* If it's a control character, we actually mean for it to be a
       * color change.
       */
      fputs(COLOR_CODES[endstring[i]], stream);

    else
      /* Otherwise, it's just regular ol' ASCII. */
      fputc(endstring[i], stream);
  }

  /* The color is already back to normal, so all that's left is to close
   * the end string with a second pipe. Then write a newline.
   */
  fputs("|\n", stream);
}

void write_byte (FILE*          stream,
                 unsigned char  byte,
                 unsigned char* color,
                 char*          endstring,
                 unsigned char* endlen)
{
  /* By default, let's assume we've been given a visible ASCII byte. The
   * color should be normal, and the pretend byte is the same as the
   * actual byte.
   */
  unsigned char new_color     = COLOR_NORMAL,
                pretend_byte  = byte;

#define SET_CHARS(X, Y, Z)  \
  if (byte X) {             \
    new_color     = Y;      \
    pretend_byte  = byte Z; \
  }

  /* This accounts for every possible non-ASCII byte. */
  SET_CHARS(      < 0x20, COLOR_CONTROL, + 0x40)
  else SET_CHARS(== 0x7f, COLOR_CONTROL, - 0x40)
  else SET_CHARS(>= 0xc0, COLOR_UTFHIGH, - 0x80)
  else SET_CHARS(>= 0x80, COLOR_UTFLOW,  - 0x40)

#undef SET_CHARS

  if (pretend_byte == 0x7f)
    /* There's a chance that this will end up being the non-ASCII DEL
     * character, in which case we want to backup another 0x40.
     */
    pretend_byte -= 0x40;

  if (new_color != *color) {
    /* If our new color is different from the previous color, we'll want
     * to output the new one.
     */
    fputs(COLOR_CODES[new_color], stream);
    endstring[(*endlen)++]  = new_color;

    /* We should also modify the previous color. */
    *color                  = new_color;
  }

  /* Whether or not we needed to print a new color, we definitely need
   * to print the hex code for the byte.
   */
  fprintf(stream, "%02x ", byte);

  /* And our visual string should contain the guaranteed-ASCII pretend
   * byte.
   */
  endstring[(*endlen)++] = pretend_byte;
}
